### PR TITLE
Add network errors to ledger funding integration test

### PR DESCRIPTION
**Pull request by snario**
**Add an end-to-end test with network errors**

This PR introduces a new end to end test file in `e2e-testing` which has networking errors built in. The hard-coded parameters in the pull request are:

- `--dropIncomingRate 0.03`: 3% chance incoming messages are dropped
-  `--dropOutgoingRate 0`: 0% chance outgoing messages are dropped
- `--meanDelay 15`: Mean 15ms delay between message send and receive (distribution)

Re-using existing utilities, the test will create 5 allocations with 5 channels each with the above failure rates during the messages exchanged between the parties. The `ensureChannelsOpen` method that was recently added is tested by this test as you can see in the logs... like this:

```
[1612888676159] DEBUG	(17644 on lhmbp13.local): Channels not yet opened, will try syncing after a timeout
    module: "PaymentManager"
    component: "ChannelPaymentManager"
    notRunningChannelIds: [
      "0xe81d7b12bf7dd64a59114a92660c7695408c20d31a9fadf6e4cf33d904e80421",
      "0xce146552223b2965311a034c0c893cedfa7bf49dc4d6999a460548acb3aaeafe",
      "0x87c204caaa7f718ad210dd58ee6f3c439fea74dd00a69613aa19ff9bc3eb16f0",
      "0x6223be1ba7fd3bba1cfb959bdc3c8bed14b59a53342c61aa81635a8d0de5441c",
      "0x069ed922b2853aa921a2ea395f6c7c54dbd587c63283b8556d44aaf85f37135e"
    ]
    retryTimeoutMs: 2000
```

This is a reasonably simple test. It creates the channels, and then attempts to make 20 payments, ignoring any errors. The goal is simply to ensure that _eventually_ if we try enough times, we will see 20 payments go through even in bad network conditions.

## Code quality
- [x] I have written clear commit messages
- [x] I have performed a self-review of my own code
- [x] I have scoped this change as narrowly as possible
- [x] I have separated logic changes from refactor changes (formatting, renames, etc.)
- [x] I have commented my code wherever necessary
- [x] I have added tests that prove my fix is effective or that my feature works
## Project management
- [x] I have applied the [appropriate labels](https://www.notion.so/Team-working-agreements-2a95c926bb5642e5a5c42e4b74a9dd24#b304e56734a74dfbb341b8b4b27b1c0c)
- [x] I have linked to relevant issues
- [x] I have added dependent tickets
- [x] I have assigned myself to this PR
- [x] I have chosen the appropriate pipeline

### CODE REVIEW

File: packages/e2e-testing/__tests__/bad-network.test.ts [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+configureEnvVariables();
+
+const LOG_FILE = `/tmp/bad-network-test-without-chain.log`;
+
+const serverArgs = [
+  `--amountOfWorkerThreads ${process.env.AMOUNT_OF_WORKER_THREADS || 0}`,
+  `--channelsPerAllocation 5`,
+  `--dropIncomingRate 0.03`, // 3% chance incoming messages are dropped
+  `--dropOutgoingRate 0.03`, // 3% chance outgoing messages are dropped
```
Comment (andrewgordstewart) [inline packages/e2e-testing/__tests__/bad-network.test.ts:36]: This will cause a flickering test until:
1.  https://github.com/statechannels/statechannels/pull/3275 is merged
2. the CM includes functionality to sync ongoing objectives.

Do we want to create an intentionally flickering tesT?

**Review by andrewgordstewart — CHANGES_REQUESTED** [review]
This looks great, except I don't think we want to create allocations in the startup script (with `numAllocations 5`).

File: packages/payments/src/channel-manager.ts [hunk 2]
```diff
@@ -295,18 +295,30 @@ export class ChannelManager implements ChannelManagementAPI {
...
+    const [resumedChannels, stillStalled] = _.partition(
+      channelResults,
+      (channelResult) => channelResult.turnNum % 2 === 1 && channelResult.turnNum >= 3
+    );
+
+    if (resumedChannels.length)
+      this.logger.debug(`Resumed stalled channels successfully`, {
+        numChannels: resumedChannels.length,
+        channels: resumedChannels.map((c) => c.channelId)
+      });
```
[the remark this answers is not in the document]
Comment (snario) [reply packages/payments/src/channel-manager.ts]: Fixed!

File: packages/e2e-testing/__tests__/bad-network.test.ts (hunk 1 above)
Comment (snario) [reply packages/e2e-testing/__tests__/bad-network.test.ts:36]: Addressed by setting `numAllocations` to `0`

File: packages/payments/src/channel-manager.ts [hunk 3]
```diff
@@ -402,7 +414,7 @@ export class ChannelManager implements ChannelManagementAPI {
 
     let results = await this.exchangeMessagesUntilOutboxIsEmpty(initialMessage);
 
-    for (const retryTimeoutMs of [2_500, 5_000, 10_000, 20_000, 40_000]) {
+    for (const retryTimeoutMs of [1_000, 2_000, 3_000, 5_000, 8_000, 13_000, 21_000]) {
```
[the remark this answers is not in the document]
Comment (snario) [reply packages/payments/src/channel-manager.ts:417]: No reason other than they were originally arbitrary and I felt like adding slightly more timeouts without increasing the total length of the call seemed reasonable based on my gut feeling about the types of scenarios that would require timeout retries.

[the remark this answers is not in the document]
Comment (andrewgordstewart) [reply packages/payments/src/channel-manager.ts:417]: Fibbonicci numbers are asymptotically exponential. `F_n+1 / F_n` quite quickly approaches ~1.6, actually, so these back-offs are essentially `range(6).map(i => 1_000 * 1.6 ** i)`.

I plan to make this configurable (eg. [here](https://github.com/statechannels/graph-payments/pull/30/commits/c842e9e8c5400755eb09fee625434e670b2a2128#diff-77cb416992eb54fd561f53ddd5bcaa1909db4e85b0a9a9d62682bf05faf9a603R186-R188)) so it can become `range(NUM).map(i => INITIAL_BACKOFF * RATIO ** i)`. This
1. allows us to test the behaviour of "retry attempts exceeded" in a sensible amount of time
2. allows the gateway to be able to decide how many retries, and how quickly

File: packages/e2e-testing/__tests__/bad-network.test.ts [hunk 4]
```diff
@@ -29,20 +30,26 @@ configureEnvVariables();
...
 const LOG_FILE = `/tmp/bad-network-test-without-chain.log`;
 
+const NUM_ALLOCATIONS = 5;
+
 const serverArgs = [
   `--amountOfWorkerThreads ${process.env.AMOUNT_OF_WORKER_THREADS || 0}`,
   `--channelsPerAllocation 5`,
   `--dropIncomingRate 0.03`, // 3% chance incoming messages are dropped
-  `--dropOutgoingRate 0.03`, // 3% chance outgoing messages are dropped
+  `--dropOutgoingRate 0`, // 3% chance outgoing messages are dropped
```
Comment (andrewgordstewart) [inline packages/e2e-testing/__tests__/bad-network.test.ts:39]: nit: this is a significant change that is not related to the commit header. ie. this is not an atomic commit.

Also, the comment is now incorrect.

File: packages/e2e-testing/__tests__/bad-network.test.ts [hunk 5]
```diff
@@ -97,18 +104,41 @@ afterAll(async () => {
...
+            ...allocation.subgraphDeploymentID,
+            display: allocation.subgraphDeploymentID.display,
+            ipfsHash: allocation.subgraphDeploymentID.ipfsHash,
+            bytes32: allocation.subgraphDeploymentID.bytes32
+          },
+          allocatedTokens: allocation.allocatedTokens,
+          createdAtEpoch: allocation.createdAtEpoch
+        },
+        num: NUM_ALLOCATIONS,
+        type: 'SetTo'
```
Comment (andrewgordstewart) [inline packages/e2e-testing/__tests__/bad-network.test.ts:128]: I thought there was a fixture for this? Let me check.

Comment (andrewgordstewart) [reply packages/e2e-testing/__tests__/bad-network.test.ts:128]: A (local) fixture is used [here](https://github.com/statechannels/graph-payments/blob/247b22941132b92fe676e594ef40328abea2ab7b/packages/payments/src/__tests__/extract-capacities.test.ts#L11).

File: packages/e2e-testing/__tests__/bad-network.test.ts [hunk 6]
```diff
@@ -118,14 +148,14 @@ describe('Payment & Receipt Managers E2E', () => {
...
-      //   logger.trace('Payment failed', {err});
-      //   continue;
-      // }
+      try {
+        await syncChannels(PAYER_SERVER_URL);
+        const {status} = await successfulPayment(PAYER_SERVER_URL);
+        if (status === 200) numPayments++;
+      } catch (err) {
+        if (err.toString() === 'Request failed with status code 500') continue;
+        throw err;
```
Comment (andrewgordstewart) [inline packages/e2e-testing/__tests__/bad-network.test.ts:157]: What's the reasoning for this change?

File: packages/e2e-testing/src/payment-server.ts [hunk 7]
```diff
@@ -304,14 +304,21 @@ function startApp(
           indexer: joi
             .object({
               url: joi.string().required(),
-              id: joi.string().required()
```
Comment (andrewgordstewart) [inline packages/e2e-testing/src/payment-server.ts:305]: Can you explain this change? Is it the case that the contracts package has changed, which necessitates a new schema for allocations?

If yes, I think that information is important to include in this git history.

**Review by andrewgordstewart — APPROVED** [review]

File: packages/e2e-testing/__tests__/bad-network.test.ts (hunk 6 above)
Comment (snario) [reply packages/e2e-testing/__tests__/bad-network.test.ts:157]: I had mistakenly thought it might be possible for a payment to fail (no free channels) but upon thinking about it more, this should not happen with this test.

File: packages/e2e-testing/src/payment-server.ts (hunk 7 above)
Comment (snario) [reply packages/e2e-testing/src/payment-server.ts:305]: Rebased with clearer commit history.

File: packages/e2e-testing/__tests__/bad-network.test.ts (hunk 5 above)
Comment (snario) [reply packages/e2e-testing/__tests__/bad-network.test.ts]: Since it is not that useful and in a different package, I'd like to opt to not re-use this for now until there is a 3rd use case warranting a generalization.
